### PR TITLE
fb: add vsync queue mechanism

### PR DESCRIPTION
**Pull request by terry0012**
## Summary
Refactor vsync queue and adapt to sim and goldfish modules.

In the new mechanism, the vendor driver does not need to create its own queue management mechanism. it is uniformly handled by the upper layer .
It also supports multiple file opening and polling for fb poll, such as polling for primary fb and overlay at the same time.

### CODE REVIEW

File: arch/sim/src/sim/sim_framebuffer.c [hunk 1]
```diff
@@ -215,6 +219,19 @@ static int sim_closewindow(struct fb_vtable_s *vtable)
   return ret;
 }
 
+#ifdef CONFIG_FB_SYNC
+static int sim_waitforvsync(struct fb_vtable_s *vtable)
```
Comment (xiaoxiang781216) [inline arch/sim/src/sim/sim_framebuffer.c:223]: let's implement in the fb driver?

**Comment by xiaoxiang781216** [comment]
@terry0012 please fix the following error:
```
/github/workspace/sources/nuttx/drivers/video/fb.c: In function 'fb_write':
Error: /github/workspace/sources/nuttx/drivers/video/fb.c:420:5: error: 'struct fb_chardev_s' has no member named 'pollready'
  420 |   fb->pollready = false;
      |     ^~
/github/workspace/sources/nuttx/drivers/video/fb.c: In function 'fb_get_planeinfo':
Warning: /github/workspace/sources/nuttx/drivers/video/fb.c:1166:57: warning: passing argument 3 of 'fb->vtable->getplaneinfo' from incompatible pointer type [-Wincompatible-pointer-types]
 1166 |   ret = fb->vtable->getplaneinfo(fb->vtable, fb->plane, &pinfo);
      |                                                         ^~~~~~
      |                                                         |
      |                                                         struct fb_planeinfo_s **
/github/workspace/sources/nuttx/drivers/video/fb.c:1166:57: note: expected 'struct fb_planeinfo_s *' but argument is of type 'struct fb_planeinfo_s **'
Error: /github/workspace/sources/nuttx/drivers/video/fb.c:1174:3: error: 'panelinfo' undeclared (first use in this function); did you mean 'mallinfo'?
 1174 |   panelinfo->fbmem  = pinfo.fbmem;
      |   ^~~~~~~~~
      |   mallinfo
/github/workspace/sources/nuttx/drivers/video/fb.c:1174:3: note: each undeclared identifier is reported only once for each function it appears in
Error: /github/workspace/sources/nuttx/drivers/video/fb.c:1174:28: error: 'pinfo' is a pointer; did you mean to use '->'?
 1174 |   panelinfo->fbmem  = pinfo.fbmem;
      |                            ^
      |                            ->
Error: /github/workspace/sources/nuttx/drivers/video/fb.c:1175:28: error: 'pinfo' is a pointer; did you mean to use '->'?
 1175 |   panelinfo->fblen  = pinfo.fblen;
      |                            ^
      |                            ->
Error: /github/workspace/sources/nuttx/drivers/video/fb.c:1176:28: error: 'pinfo' is a pointer; did you mean to use '->'?
 1176 |   panelinfo->bpp    = pinfo.bpp;
      |                            ^
      |   
```

**Comment by pkarashchenko** [comment]
Please fix CI build

File: arch/sim/src/sim/sim_framebuffer.c (hunk 1 above)
Comment (terry0012) [reply arch/sim/src/sim/sim_framebuffer.c]: OK,  we remove it first and add it to fb driver in the next commit.

**Comment by terry0012** [comment]
> @terry0012 please fix the following error:
> 
> ```
> /github/workspace/sources/nuttx/drivers/video/fb.c: In function 'fb_write':
> Error: /github/workspace/sources/nuttx/drivers/video/fb.c:420:5: error: 'struct fb_chardev_s' has no member named 'pollready'
>   420 |   fb->pollready = false;
>       |     ^~
> /github/workspace/sources/nuttx/drivers/video/fb.c: In function 'fb_get_planeinfo':
> Warning: /github/workspace/sources/nuttx/drivers/video/fb.c:1166:57: warning: passing argument 3 of 'fb->vtable->getplaneinfo' from incompatible pointer type [-Wincompatible-pointer-types]
>  1166 |   ret = fb->vtable->getplaneinfo(fb->vtable, fb->plane, &pinfo);
>       |                                                         ^~~~~~
>       |                                                         |
>       |                                                         struct fb_planeinfo_s **
> /github/workspace/sources/nuttx/drivers/video/fb.c:1166:57: note: expected 'struct fb_planeinfo_s *' but argument is of type 'struct fb_planeinfo_s **'
> Error: /github/workspace/sources/nuttx/drivers/video/fb.c:1174:3: error: 'panelinfo' undeclared (first use in this function); did you mean 'mallinfo'?
>  1174 |   panelinfo->fbmem  = pinfo.fbmem;
>       |   ^~~~~~~~~
>       |   mallinfo
> /github/workspace/sources/nuttx/drivers/video/fb.c:1174:3: note: each undeclared identifier is reported only once for each function it appears in
> Error: /github/workspace/sources/nuttx/drivers/video/fb.c:1174:28: error: 'pinfo' is a pointer; did you mean to use '->'?
>  1174 |   panelinfo->fbmem  = pinfo.fbmem;
>       |                            ^
>       |                            ->
> Error: /github/workspace/sources/nuttx/drivers/video/fb.c:1175:28: error: 'pinfo' is a pointer; did you mean to use '->'?
>  1175 |   panelinfo->fblen  = pinfo.fblen;
>       |                            ^
>       |                            ->
> Error: /github/workspace/sources/nuttx/drivers/video/fb.c:1176:28: error: 'pinfo' is a pointer; did you mean to use '->'?
>  1176 |   panelinfo->bpp    = pinfo.bpp;
>       |                            ^
>       |   
> ```

got it.

**Comment by terry0012** [comment]
The following errors are caused by the naming conventions of X11 and can probably be ignored.

`Error: /home/runner/work/nuttx/nuttx/nuttx/arch/sim/src/sim/posix/sim_x11framebuffer.c:109:2: error: Mixed case identifier found
Error: /home/runner/work/nuttx/nuttx/nuttx/arch/sim/src/sim/posix/sim_x11framebuffer.c:110:2: error: Mixed case identifier found
Error: Process completed with exit code 1.`

**Comment by xiaoxiang781216** [comment]
> The following errors are caused by the naming conventions of X11 and can probably be ignored.
> 
> `Error: /home/runner/work/nuttx/nuttx/nuttx/arch/sim/src/sim/posix/sim_x11framebuffer.c:109:2: error: Mixed case identifier found Error: /home/runner/work/nuttx/nuttx/nuttx/arch/sim/src/sim/posix/sim_x11framebuffer.c:110:2: error: Mixed case identifier found Error: Process completed with exit code 1.`

let's ignore "Mixed case identifier" warning, but fix the compiler error:
```
Error: sim/sim_lcd.c:468:9: error: too few arguments to function 'sim_x11initialize'
  468 |   ret = sim_x11initialize(CONFIG_SIM_FBWIDTH, CONFIG_SIM_FBHEIGHT,
      |         ^~~~~~~~~~~~~~~~~
```

File: arch/sim/Kconfig [hunk 2]
```diff
@@ -355,6 +355,13 @@ config SIM_FBBPP
 		If you use the X11 display emulation, the selected BPP must match the BPP
 		of your graphics hardware (probably 32 bits).  Default: 8
 
+config SIM_FBCOUNT
```
Comment (xiaoxiang781216) [inline arch/sim/Kconfig:358]: change to SIM_FRAMEBUFFER_COUNT and depends on SIM_FRAMEBUFFER

Comment (terry0012) [reply arch/sim/Kconfig]: Done.

**Review by xiaoxiang781216 — APPROVED** [review]

File: arch/sim/src/sim/sim_framebuffer.c [hunk 3]
```diff
@@ -455,9 +462,12 @@ int up_fbinitialize(int display)
   int ret = OK;
 
 #ifdef CONFIG_SIM_X11FB
+  g_planeinfo.xres_virtual = CONFIG_SIM_FBWIDTH;
+  g_planeinfo.yres_virtual = CONFIG_SIM_FBHEIGHT * CONFIG_SIM_FRAMEBUFFER_COUNT;
```
Comment (pkarashchenko) [inline arch/sim/src/sim/sim_framebuffer.c:466]: Good to fix this

File: drivers/video/fb.c [hunk 4]
```diff
@@ -41,17 +41,30 @@
...
  ****************************************************************************/
 
+struct fb_priv_s
+{
+  int overlay;                    /* Overlay number */
+};
+
+struct fb_paninfo_s
+{
+  FAR struct circbuf_s buf;       /* Pan buffer queued list */
```
[the remark this answers is not in the document]
Comment (xiaoxiang781216) [reply drivers/video/fb.c:61]: Fixed here: https://github.com/apache/nuttx/pull/10648

**Pull request merged**